cpp/src/Ice/ObjectAdapterI.cpp

```diff
@@ -373,6 +373,20 @@ Ice::ObjectAdapterI::destroy() noexcept
         _reference = 0;
         _objectAdapterFactory = 0;
 
+#if defined(ICE_USE_SCHANNEL)
```

Comment:
I would just put this code in the destructor and remove the const cast clearing.

All the cleanup before is to break cycles (or at least it should be - some of it seems over zealous).